activesupport/lib/active_support/tagged_logging.rb

```diff
@@ -71,11 +69,7 @@ def current_tags
 
       def tags_text
         tags = current_tags
-        if tags.one?
-          +"[#{tags[0]}] "
-        elsif tags.any?
-          tags.collect { |tag| "[#{tag}] " }.join
-        end
+        "[#{tags.join("] [")}] " unless tags.empty?
```

Comment:
I wonder if it would make sense to cache `tags_text`? Assuming more than one message is logged with the same tags we'd save a bit right?

Reply:
It might.  Since tags can be pushed / popped whenever, we would need to keep the memoized tag string in sync with the `current_tags` array.  Either by resetting the memoized tag string on each push / pop; or by concatting onto the string when pushing, and using the lengths of tags when popping to truncate the string; or by concatting onto the string, and pushing the string length instead of the tag, so no arithmetic is necessary when popping.

Also, since tags are per-thread, the memoized tag string would need to be in a thread variable (not an ivar).  I did not want to improve the performance of individual logging calls at the cost of making the block form of `tagged` slower.  Though I have not yet benchmarked that tradeoff because:  An idea that I was playing around with is a `TagStack` helper class.  It would contain all the optimized logic of pushing / popping and concatenation.  And it could be stored in a single thread variable.  But I wanted to improve the baseline performance before doing any comparisons.

Reply:
> TagStack helper class. It would contain all the optimized logic of pushing / popping and concatenation. And it could be stored in a single thread variable.

That was what I had in mind too :)

Reply:
I tried the approaches that I mentioned, as well as a few more variations.  In the end, simply resetting the memoized string was the fastest approach.

It would be nice if Ruby provided simple / fast methods for truncating a string *without* returning (i.e. allocating) the removed portion.  The closest that I know of is `StringIO#seek`, which I did try in a few of the variations, but the overhead was too much.